ethexe/runtime/common/src/journal.rs

```diff
@@ -340,39 +294,48 @@ impl<S: Storage> JournalHandler for Handler<'_, S> {
         program_id: ProgramId,
         pages_data: BTreeMap<GearPage, PageBuf>,
     ) {
-        let mut new_page_hashes = BTreeMap::new();
-        for (page, data) in pages_data {
-            let hash = self.storage.write_page_data(data);
-            new_page_hashes.insert(page, hash);
-        }
-
-        self.update_program(program_id, |state, storage| {
-            let state::Program::Active(mut active_state) = state.program else {
-                return None;
+        self.update_state_with_storage(program_id, |storage, state| {
+            let Program::Active(ActiveProgram {
+                ref mut pages_hash, ..
+            }) = state.program
+            else {
+                anyhow::bail!("an attempt to update pages data of inactive program");
             };
 
-            let mut pages_map = active_state.pages_hash.with_hash_or_default(|hash| {
-                storage.read_pages(hash).expect("Failed to read pages")
-            });
+            let new_pages = storage.store_pages(pages_data);
 
-            for (page, hash) in new_page_hashes {
-                pages_map.insert(page, hash);
-            }
-
-            let changed_active_state = state::ActiveProgram {
-                pages_hash: storage.write_pages(pages_map).into(),
-                ..active_state
-            };
+            *pages_hash = storage.modify_memory_pages(pages_hash.clone(), |pages| {
+                for (page, data) in new_pages {
+                    pages.insert(page, data);
+                }
+            })?;
 
-            Some(ProgramState {
-                program: state::Program::Active(changed_active_state),
-                ..state
-            })
+            Ok(())
         });
     }
 
-    fn update_allocations(&mut self, program_id: ProgramId, allocations: IntervalsTree<WasmPage>) {
-        todo!()
+    fn update_allocations(
+        &mut self,
+        program_id: ProgramId,
+        new_allocations: IntervalsTree<WasmPage>,
+    ) {
+        self.update_state_with_storage(program_id, |storage, state| {
+            let Program::Active(ActiveProgram {
+                ref mut allocations_hash,
+                ..
+            }) = state.program
+            else {
+                anyhow::bail!("an attempt to update allocations of inactive program");
+            };
+
+            // TODO (breathx): remove data for difference pages.
```

Comment:
this should not be done here because some other programs may have pages with the same data

Reply:
It should be, because we have pages hash as has of map page -> page hash, where pages.all(|p| allocations.contains(p))

Reply:
Data for (shared) pages won't be deleted from storage

Reply:
So, it's not data it's pages hashes from pages map. Ok